src/coreComponents/functions/MultivariableTableFunctionKernels.hpp

```diff
@@ -0,0 +1,326 @@
+      {
+        printf( "Interpolation warning: axis is out of limits (%lf; %lf) with value %lf, extrapolation is applied\n", axisMin, axisMax, axisCoordinate );
+      }
+    }
+    else if( axisIntervalIndex > (axisPoints - 2))
+    {
+      axisIntervalIndex = axisPoints - 2;
+      if( axisCoordinate > axisMax )
+      {
+        printf( "Interpolation warning: axis is out of limits (%lf; %lf) with value %lf, extrapolation is applied\n", axisMin, axisMax, axisCoordinate );
+      }
+    }
+
+    axisLow = axisIntervalIndex * axisStep + axisMin;
+    axisMult = (axisCoordinate - axisLow) * axisStepInv;
+    return axisIntervalIndex;
+  }
+
+
+  /**
+   * @brief interpolate all operator values and derivatives at a given point
+   *
+   * @param[in] axisCoordinates coordinates of a point
+   * @param[in] hypercubeData data of target hypercube
+   * @param[in] axisLows array of left coordinates of target axis intervals
+   * @param[in] axisMults array of weights of right coordinates of target axis intervals
+   * @param[in] axisStepInvs array of inversions of axis steps
+   * @param[out] values interpolated operator values
+   * @param[out] derivatives derivatives of interpolated operators
+   */
+  GEOSX_HOST_DEVICE
+  inline
+  void
+  interpolatePointWithDerivatives( real64 const * const axisCoordinates,
+                                   real64 const * const hypercubeData,
+                                   real64 const * const axisLows,
+                                   real64 const * const axisMults,
+                                   real64 const * const axisStepInvs,
+                                   real64 *values,
+                                   real64 *derivatives ) const
+  {
+    integer pwr = numVerts / 2;   // distance between high and low values
+    real64 workspace[2 * numVerts - 1][numOps];
+
+    // copy operator values for all vertices
+    for( integer i = 0; i < numVerts; ++i )
+    {
+      for( integer j = 0; j < numOps; ++j )
+      {
+        workspace[i][j] = hypercubeData[i * numOps + j];
+      }
+    }
+
+    for( integer i = 0; i < numDims; ++i )
+    {
+
+      for( integer j = 0; j < pwr; ++j )
+      {
+        for( integer op = 0; op < numOps; ++op )
+        {
+          // update own derivative
+          workspace[2 * numVerts - (numVerts >> i) + j][op] = (workspace[j + pwr][op] - workspace[j][op]) * axisStepInvs[i];
+        }
+
+        // update all dependent derivatives
+        for( integer k = 0; k < i; k++ )
+        {
+          for( integer op = 0; op < numOps; ++op )
+          {
+            workspace[2 * numVerts - (numVerts >> k) + j][op] = workspace[2 * numVerts - (numVerts >> k) + j][op] + axisMults[i] *
+                                                                (workspace[2 * numVerts - (numVerts >> k) + j + pwr][op] -
+                                                                 workspace[2 * numVerts - (numVerts >> k) + j][op]);
+          }
+        }
```

Comment:
The indexing looks pretty fancy! For my own understanding, what is a "dependent derivative" here (maybe give an example)? Why do you need the second loop to go from 0 to `i`? (I am not suggest to change anything here, it looks good, I am just trying to understand the logic)

Reply:
Yeah, nice question! I've implemented this code based on the paper http://dx.doi.org/10.1090/S0025-5718-1988-0917826-0. Basically, you need to do a sort of reduction to compute both values and derivatives (you can look at the Figure 1 from the paper). Imagine, we have a 2D space and hence 4 vertices. We choose to interpolate first along axis X, and then - along Y. When we interpolate along X, we "reduce" 4 vertices into 2, which we then use to interpolate along Y. However, the result of the second interpolation along Y will influence on the derivative along X (or in general, all previous axes), so those derivatives I call dependent, and they need to be updated in the second loop.